src/main/java/jenkins/scm/api/SCMEvent.java

```diff
@@ -247,14 +309,83 @@ public int hashCode() {
     @Override
     public String toString() {
         return String.format(
-                "SCMEvent{type=%s, timestamp=%tc, payload=%s}",
+                "SCMEvent{type=%s, timestamp=%tc, payload=%s, origin=%s}",
                 type,
                 timestamp,
-                payload
+                payload,
+                StringUtils.defaultIfBlank(origin, ORIGIN_UNKNOWN)
         );
     }
 
     /**
+     * Helper method to get the origin of an event from a {@link HttpServletRequest}. The current format is the
+     * list of hostname / ip addresses from the request (parsing {@code X-Forwarded-For} headers) separated by
+     * {@code →} followed by a {@code ⇒} and finally the requested URL (omitting the query portion of the URL).
+     *
+     * @param req the {@link HttpServletRequest} or {@code null} (this is to allow passing
+     *            {@link Stapler#getCurrentRequest()} without having to check for {@code null})
+     * @return the origin of the event or {@code null} if the {@link HttpServletRequest} is null.
+     * @since 2.0.3
+     */
+    @CheckForNull
+    public static String originOf(@CheckForNull HttpServletRequest req) {
+        if (req == null) {
+            return null;
+        }
+        String last = null;
+        StringBuilder result = new StringBuilder();
+        // TODO RFC 7239 support once standard is approved
+        String header = req.getHeader("X-Forwarded-For");
+        if (StringUtils.isNotBlank(header)) {
```

Comment:
it could be replaced by a utility from StringUtils *except* we need to catch the last value in case it matches the remote host/addr